src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.Structs.cs

```diff
@@ -0,0 +1,164 @@
+{
+    internal sealed partial class NullableWalker
+    {
+        /// <summary>
+        /// Represents the result of visiting an expression.
+        /// Contains a result type which tells us whether the expression may be null,
+        /// and an l-value type which tells us whether we can assign null to the expression.
+        /// </summary>
+        [DebuggerDisplay("{GetDebuggerDisplay(), nq}")]
+        private readonly struct VisitResult
+        {
+            public readonly TypeWithState RValueType;
+            public readonly TypeWithAnnotations LValueType;
+
+            public VisitResult(TypeWithState rValueType, TypeWithAnnotations lValueType)
+            {
+                RValueType = rValueType;
+                LValueType = lValueType;
+                // https://github.com/dotnet/roslyn/issues/34993: Doesn't hold true for Tuple_Assignment_10. See if we can make it hold true
+                //Debug.Assert((RValueType.Type is null && LValueType.TypeSymbol is null) ||
+                //             RValueType.Type.Equals(LValueType.TypeSymbol, TypeCompareKind.ConsiderEverything | TypeCompareKind.AllIgnoreOptions));
+            }
+
+            public VisitResult(TypeSymbol type, NullableAnnotation annotation, NullableFlowState state)
+            {
+                RValueType = TypeWithState.Create(type, state);
+                LValueType = TypeWithAnnotations.Create(type, annotation);
+                Debug.Assert(RValueType.Type.Equals(LValueType.Type, TypeCompareKind.ConsiderEverything));
+            }
+
+            internal string GetDebuggerDisplay() => $"{{LValue: {LValueType.GetDebuggerDisplay()}, RValue: {RValueType.GetDebuggerDisplay()}}}";
+        }
+
+        /// <summary>
+        /// Represents the result of visiting an argument expression.
+        /// In addition to storing the <see cref="VisitResult"/>, also stores the <see cref="LocalState"/>
+        /// for reanalyzing a lambda.
+        /// </summary>
+        [DebuggerDisplay("{VisitResult.GetDebuggerDisplay(), nq}")]
+        private readonly struct VisitArgumentResult
+        {
+            public readonly VisitResult VisitResult;
+            public readonly Optional<LocalState> StateForLambda;
+            public TypeWithState RValueType => VisitResult.RValueType;
+            public TypeWithAnnotations LValueType => VisitResult.LValueType;
+
+            public VisitArgumentResult(VisitResult visitResult, Optional<LocalState> stateForLambda)
+            {
+                VisitResult = visitResult;
+                StateForLambda = stateForLambda;
+            }
+        }
+
+        /// <summary>
+        /// Contains a checkpoint for the NullableWalker at any given point of execution, used for restoring the walker to
+        /// a specific point for speculatively analyzing a piece of code that does not appear in the original tree.
+        /// </summary>
+        internal readonly struct Checkpoint
```

Comment:
>Checkpoint [](start = 33, length = 10)

Should this be named "Snapshot" perhaps? "Checkpoint" sounds like a point in time where a number of checks are performed rather than capturing state.

Reply:
I don't know if it's the gamer in me, but in video games a "checkpoint" is the last state you go back to if you fail the level. I can use a different word if you really want, but checkpoint is correct for this as well.

---
In reply to: [286230409](https://github.com/dotnet/roslyn/pull/35850#discussion_r286230409) [](ancestors = 286230409)